internal/rukpak/source/containers_image_test.go

```diff
@@ -0,0 +1,397 @@
+		BaseCachePath: t.TempDir(),
+		SourceContext: buildPullContext(t, imageTagRef),
+	}
+	bundleSource := &source.BundleSource{
+		Name: "test-bundle",
+		Type: source.SourceTypeImage,
+		Image: &source.ImageSource{
+			Ref: reference.TrimNamed(imageTagRef).String(),
+		},
+	}
+
+	// Attempt to pull and unpack the image
+	_, err := unpacker.Unpack(context.Background(), bundleSource)
+	// Check Result
+	require.ErrorContains(t, err, "tag or digest is needed")
+}
+
+func TestUnpackUnservedTaggedImageReference(t *testing.T) {
+	imageTagRef, _, cleanup := setupRegistry(t)
+	defer cleanup()
+
+	unpacker := &source.ContainersImageRegistry{
+		BaseCachePath: t.TempDir(),
+		SourceContext: buildPullContext(t, imageTagRef),
+	}
+	bundleSource := &source.BundleSource{
+		Name: "test-bundle",
+		Type: source.SourceTypeImage,
+		Image: &source.ImageSource{
+			// Use a valid reference that is not served
+			Ref: fmt.Sprintf("%s:unserved-tag", reference.TrimNamed(imageTagRef)),
+		},
+	}
+
+	// Attempt to pull and unpack the image
+	_, err := unpacker.Unpack(context.Background(), bundleSource)
+	// Check Result
+	require.ErrorContains(t, err, "manifest unknown")
+}
+
+func TestUnpackUnservedCanonicalImageReference(t *testing.T) {
+	imageTagRef, imageDigestRef, cleanup := setupRegistry(t)
+	defer cleanup()
+
+	unpacker := &source.ContainersImageRegistry{
+		BaseCachePath: t.TempDir(),
+		SourceContext: buildPullContext(t, imageTagRef),
+	}
+
+	origRef := imageDigestRef.String()
+	nonExistentRef := origRef[:len(origRef)-1] + "1"
+	bundleSource := &source.BundleSource{
+		Name: "test-bundle",
+		Type: source.SourceTypeImage,
+		Image: &source.ImageSource{
+			// Use a valid reference that is not served
+			Ref: nonExistentRef,
+		},
+	}
+
+	// Attempt to pull and unpack the image
+	_, err := unpacker.Unpack(context.Background(), bundleSource)
+	// Check Result
+	require.ErrorContains(t, err, "manifest unknown")
+}
+
+func TestUnpackInvalidSourceType(t *testing.T) {
+	unpacker := &source.ContainersImageRegistry{}
+	// Create BundleSource with invalid source type
+	bundleSource := &source.BundleSource{
+		Type: "invalid",
+	}
+
+	shouldPanic := func() {
+		// Attempt to pull and unpack the image
+		_, err := unpacker.Unpack(context.Background(), bundleSource)
+		if err != nil {
+			t.Error("func should have panicked")
+		}
+	}
+	require.Panics(t, shouldPanic)
+}
+
+func TestUnpackInvalidNilImage(t *testing.T) {
+	unpacker := &source.ContainersImageRegistry{
+		BaseCachePath: t.TempDir(),
+	}
+	// Create BundleSource with nil Image
+	bundleSource := &source.BundleSource{
+		Name:  "test-bundle",
+		Type:  source.SourceTypeImage,
+		Image: nil,
+	}
+
+	// Attempt to unpack
+	result, err := unpacker.Unpack(context.Background(), bundleSource)
+	require.ErrorContains(t, err, "nil image source")
+	require.NoDirExists(t, filepath.Join(unpacker.BaseCachePath, bundleSource.Name))
+	assert.Nil(t, result)
+}
+
+func TestUnpackInvalidImageRef(t *testing.T) {
```

Comment:
Should we ensure the error type returned is unrecoverable?

Reply:
[f21e3e5003b70b5b14eb73fc3c5170dd4f147425](https://github.com/operator-framework/operator-controller/compare/df384cd6cb07fbd53f8d037b94d24a96a60c50b9..f21e3e5003b70b5b14eb73fc3c5170dd4f147425)